### PR TITLE
Compiler warning fixes and improvement

### PR DESCRIPTION
**Pull request by kupsch**
Update the code to eliminate warnings mainly by clang, and a few by gcc. Significant items include:

- Safely perform memory access of data types where the pointer may not be properly aligned. Previously these just cast and accessed the memory which would have caused a fault on processors that do not allow unaligned memory access. On systems that allow unaligned access this will be as efficient as before and on those that do not it will work correctly.
- Replaced code that is deprecated in newer versions of C++ and C.
- Builds without warnings using gcc 6-13 and clang 7-15 for all 4 build types, and C++ language standards of 11, 14, 17, 20 and 23 with the exception of gcc-12 MINSIZEREL (warning), and clang-14 with C++ 20 and 23 (error)
- Many other small cleanups

### CODE REVIEW

**Review by hainest — CHANGES_REQUESTED** [review]
dyninstAPI/src/codegen-x86.C:1438: It looks like this could be a `append_memory_as(newInsn, *origInsn++);`

All of the BPatch classes are part of the public API but don't attempt to correctly implement the rule of 0/5 and many have pointer members. I guess we should put that on our (ever-growing) list of things to clean up?

>warning fix:  missing noexcept

Why did the compilers recommend this? It's not clear to me why these would need to be noexcept.

File: dyninstAPI/src/codegen-x86.C [hunk 1]
```diff
@@ -1424,13 +1410,9 @@ bool insnCodeGen::modifyDisp(signed long newDisp, instruction &insn, codeGen &ge
     if (origDisp != newDisp) {
         // Replace displacement
         if (is_disp8(newDisp)) {
-            *((signed char *)newInsn) = (signed char)(newDisp);
-            newInsn += sizeof(signed char);
-            newInsnSz += sizeof(signed char);
```
[the remark this answers is not in the document]
Comment (kupsch) [reply dyninstAPI/src/codegen-x86.C:1429]: @hainest, this is fixed along with additional append_memory_as_byte changes in the 2 additional commits to this PR.  Please review.

File: dyninstAPI/src/codegen-x86.C [hunk 2]
```diff
@@ -284,29 +279,26 @@ void insnCodeGen::generatePush64(codeGen &gen, Address val)
...
   for (int i = 3; i >= 0; i--) {
-    unsigned short word = static_cast<unsigned short>((val >> (16 * i)) & 0xffff);
-    *insn++ = 0x66; // operand size override
-    *insn++ = 0x68; // push immediate (16-bits b/c of prefix)
-    *(unsigned short *)insn = word;
-    insn += 2;
+    uint16_t word = static_cast<unsigned short>((val >> (16 * i)) & 0xffff);
+    append_memory_as_byte(insn, 0x66); // operand size override
+    append_memory_as_byte(insn, 0x68); // push immediate (16-bits b/c of prefix)
+    append_memory_as(insn, uint32_t{word));
```
Comment (hainest) [inline dyninstAPI/src/codegen-x86.C:285]: Unbalanced braces:   uint32_t{word) -> uint32_t{word}

**Review by hainest — CHANGES_REQUESTED** [review]
Just the one item. Looks good, otherwise.

**Comment by kupsch** [comment]
The noexcept warnings come from gcc-12 with C++20, construct_at has noexcept specifier that evaluates to false, but could be true if these methods were noexcept.  One of the warnings is:

```
In file included from /usr/include/c++/12/bits/stl_iterator.h:85,
                 from /usr/include/c++/12/bits/stl_algobase.h:67,
                 from /usr/include/c++/12/list:60,
                 from /nobackup/kupsch/kupsch/dyninst/symtabAPI/src/LineInformation.C:32:
/usr/include/c++/12/bits/stl_construct.h: In instantiation of ‘constexpr decltype (::new(void*(0)) _Tp) std::construct_at(_Tp*, _Args&& ...) [with _Tp = Dyninst::SymtabAPI::AddressRange; _Args = {Dyninst::SymtabAPI::AddressRange}; decltype (::new(void*(0)) _Tp) = Dyninst::SymtabAPI::AddressRange*]’:
/usr/include/c++/12/bits/alloc_traits.h:518:21:   required from ‘static constexpr void std::allocator_traits<std::allocator<_Up> >::construct(allocator_type&, _Up*, _Args&& ...) [with _Up = Dyninst::SymtabAPI::AddressRange; _Args = {Dyninst::SymtabAPI::AddressRange}; _Tp = Dyninst::SymtabAPI::AddressRange; allocator_type = std::allocator<Dyninst::SymtabAPI::AddressRange>]’
/usr/include/c++/12/bits/vector.tcc:117:30:   required from ‘constexpr std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::emplace_back(_Args&& ...) [with _Args = {Dyninst::SymtabAPI::AddressRange}; _Tp = Dyninst::SymtabAPI::AddressRange; _Alloc = std::allocator<Dyninst::SymtabAPI::AddressRange>; reference = Dyninst::SymtabAPI::AddressRange&]’
/usr/include/c++/12/bits/stl_vector.h:1294:21:   required from ‘constexpr void std::vector<_Tp, _Alloc>::push_back(value_type&&) [with _Tp = Dyninst::SymtabAPI::AddressRange; _Alloc = std::allocator<Dyninst::SymtabAPI::AddressRange>; value_type = Dyninst::SymtabAPI::AddressRange]’
/nobackup/kupsch/kupsch/dyninst/symtabAPI/src/LineInformation.C:152:25:   required from here
/usr/include/c++/12/bits/stl_construct.h:95:14: warning: noexcept-expression evaluates to ‘false’ because of a call to ‘Dyninst::SymtabAPI::AddressRange::AddressRange(const Dyninst::SymtabAPI::AddressRange&)’ [-Wnoexcept]
   95 |     noexcept(noexcept(::new((void*)0) _Tp(std::declval<_Args>()...)))
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /nobackup/kupsch/kupsch/dyninst/symtabAPI/h/Module.h:44,
                 from /nobackup/kupsch/kupsch/dyninst/symtabAPI/src/LineInformation.C:37:
/nobackup/kupsch/kupsch/dyninst/symtabAPI/h/RangeLookup.h:69:13: note: but ‘Dyninst::SymtabAPI::AddressRange::AddressRange(const Dyninst::SymtabAPI::AddressRange&)’ does not throw; perhaps it should be declared ‘noexcept’
   69 |             AddressRange(const AddressRange& other) : std::pair<Offset, Offset>(other)
      |             ^~~~~~~~~~~~
```

**Review by hainest — APPROVED** [review]

**Comment by hainest** [comment]
@kupsch Looks good.

**Pull request merged**